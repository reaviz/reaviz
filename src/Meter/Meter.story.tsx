import React from 'react';
import { Meter } from './Meter';

export default {
  tags: ['snapshot'],
  title: 'Charts/Meter',
  component: Meter
};

export const Basic = () => <Meter value={50} style={{ width: 300 }} />;

export const MinMax = () => (
  <Meter value={40} style={{ width: 300 }} min={0} max={50} />
);
