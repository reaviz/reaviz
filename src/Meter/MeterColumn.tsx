import React, { FC } from 'react';
import { motion } from 'motion/react';
import { DEFAULT_TRANSITION, schemes } from '@/common';

export interface MeterColumnProps {
  /**
   * The value of the meter.
   */
  value: number;

  /**
   * The height of the column.
   */
  height?: number;

  /**
   * The index of the column.
   */
  index: number;

  /**
   * The scale to use.
   */
  scale: any;

  /**
   * The active fill color.
   */
  activeFill: string;

  /**
   * The inactive fill color.
   */
  inActiveFill: string;

  /**
   * Additional class names to apply.
   */
  className?: string;

  /**
   * The total number of columns. Set internally.
   */
  count?: number;

  /**
   * Whether to animate the column.
   */
  animated?: boolean;
}

export const MeterColumn: FC<Partial<MeterColumnProps>> = ({
  index,
  scale,
  value,
  count,
  height = 32,
  className,
  animated = true,
  activeFill = schemes.cybertron[0],
  inActiveFill = '#414242'
}) => {
  const isActive = scale(index) <= scale(value);
  const fill = isActive ? activeFill : inActiveFill;
  const transition = animated
    ? {
        ...DEFAULT_TRANSITION,
        delay: (index / count) * 0.5
      }
    : {
        type: false as const,
        delay: 0
      };

  return (
    <motion.div
      key={index}
      initial={{
        background: inActiveFill,
        height
      }}
      animate={{
        background: fill,
        height
      }}
      transition={transition}
      className={className}
    />
  );
};
