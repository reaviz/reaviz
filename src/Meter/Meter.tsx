import React, { FC, ReactElement } from 'react';
import { scaleLinear } from 'd3-scale';
import classNames from 'classnames';
import { MeterColumn, MeterColumnProps } from './MeterColumn';
import { CloneElement } from 'reablocks';
import css from './Meter.module.css';

export interface MeterProps {
  /**
   * The value of the meter.
   */
  value: number;

  /**
   * The minimum value of the meter.
   */
  min?: number;

  /**
   * The maximum value of the meter.
   */
  max?: number;

  /**
   * The number of columns to display.
   */
  columns?: number;

  /**
   * Additional class names to apply.
   */
  className?: string;

  /**
   * The gap between columns.
   */
  gap?: number;

  /**
   * Additional styles to apply.
   */
  style?: React.CSSProperties;

  /**
   * The column to render.
   */
  column: ReactElement<MeterColumnProps, typeof MeterColumn> | null;
}

export const Meter: FC<Partial<MeterProps>> = ({
  min = 0,
  max = 100,
  className,
  column = <MeterColumn />,
  gap = 15,
  style = {},
  value,
  columns = 10
}) => {
  const scale = scaleLinear().domain([min, max]).range([0, 100]);
  const cols = scale.ticks(columns);

  return (
    <div
      className={classNames(css.container, className)}
      style={{ gap: `${gap}px`, ...style }}
    >
      {cols.map((index) => (
        <CloneElement<MeterColumnProps>
          element={column}
          key={index}
          count={cols.length}
          index={index}
          scale={scale}
          value={value}
        />
      ))}
    </div>
  );
};
