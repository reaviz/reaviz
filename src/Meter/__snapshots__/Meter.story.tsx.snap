// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts/Meter Basic smoke-test 1`] = `
<div class="_container_13giw_1"
     style="gap: 15px; width: 300px;"
>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(65, 66, 66); height: 32px;">
  </div>
  <div style="background: rgb(65, 66, 66); height: 32px;">
  </div>
  <div style="background: rgb(65, 66, 66); height: 32px;">
  </div>
  <div style="background: rgb(65, 66, 66); height: 32px;">
  </div>
  <div style="background: rgb(65, 66, 66); height: 32px;">
  </div>
</div>
`;

exports[`Charts/Meter MinMax smoke-test 1`] = `
<div class="_container_13giw_1"
     style="gap: 15px; width: 300px;"
>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(45, 96, 232); height: 32px;">
  </div>
  <div style="background: rgb(65, 66, 66); height: 32px;">
  </div>
  <div style="background: rgb(65, 66, 66); height: 32px;">
  </div>
</div>
`;
